File: 02_wizards/csharpdotnet/Wizards/Wizard.cs
namespace Wizards
{
    #region

    using System.Collections.Generic;

    #endregion

    public class Wizard
    {
        private readonly SpellSchool specialization;
        private readonly IList<Spell> spellBook;

        public Wizard(SpellSchool school)
        {
            this.specialization = school;
            this.spellBook = new List<Spell>();
        }

        public string Cast(Spell spell)
        {
            if (!spellBook.Contains(spell))
            {
                return "fumble";
            }

            return AmISpecializedIn(spell) ? "Effective " + spell.Name : spell.Name;
        }

        private bool AmISpecializedIn(Spell spell)
        {
            return specialization.Equals(spell.School);
        }

        public void Memorize(Spell spell)
        {
            spellBook.Add(spell);
        }
    }
}

File: 02_wizards/java/src/main/java/wizards/Wizard.java
package wizards;

import java.util.ArrayList;
import java.util.List;

public abstract class Wizard {
	private SpellSchool spellSchool;

	protected Wizard(SpellSchool spellSchool) {
		this.spellSchool = spellSchool;
	}

	private List<Spell> spellBook = new ArrayList<Spell>();

	public void memorize(Spell spell) {
		this.spellBook.add(spell);
	}

	public String cast(Spell spell) {
		if (this.spellBook.contains(spell)) {
			return (spell.getSpellSchool() == this.spellSchool ? "Effective " : "") + spell.getName();
		}
		
		return "Fumble";
	}
}
